cdc/processor/sinkmanager/worker_impl.go

```diff
@@ -0,0 +1,215 @@
+	// enableOldValue indicates whether to enable the old value feature.
+	// If it is enabled, we need to deal with the compatibility of the data format.
+	enableOldValue bool
+}
+
+// newWorker creates a new worker.
+// nolint:deadcode
+func newWorker(
+	changefeedID model.ChangeFeedID,
+	sortEngine sorter.EventSortEngine,
+	redoManager redo.LogManager,
+	quota *memQuota,
+	splitTxn bool,
+	enableOldValue bool,
+) worker {
+	return &workerImpl{
+		changefeedID:   changefeedID,
+		sortEngine:     sortEngine,
+		redoManager:    redoManager,
+		memQuota:       quota,
+		splitTxn:       splitTxn,
+		enableOldValue: enableOldValue,
+	}
+}
+
+func (w *workerImpl) receiveTableSinkTask(ctx context.Context, taskChan <-chan *tableSinkTask) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case task := <-taskChan:
+			// First time to run the task, we have initialized memory quota for the table.
+			availableMem := defaultRequestMemSize
+			events := make([]*model.PolymorphicEvent, 0, 1024)
+
+			// Used to record the last written position.
+			// We need to use it to update the lower bound of the table sink.
+			var lastPos sorter.Position
+			currentTotalSize := uint64(0)
+			batchID := uint64(1)
+			// We have to get the latest barrier ts, this is because the barrier ts may be updated.
+			// We always want to get the latest value.
+			currentBarrierTs := task.upperBarrierTs.Load()
+			upperBound := sorter.Position{
+				StartTs:  currentBarrierTs - 1,
+				CommitTs: currentBarrierTs,
+			}
+
+			// Two functions to simplify the code.
+			// It captures some variables in the outer scope.
+			appendEventsAndRecordCurrentSize := func() error {
+				size, err := w.appendEventsToTableSink(task, events)
+				if err != nil {
+					return errors.Trace(err)
+				}
+				currentTotalSize += size
+				events = events[:0]
+				return nil
+			}
+			advanceTableSinkAndResetCurrentSize := func(event *model.PolymorphicEvent) error {
+				err := w.advanceTableSink(task, event.CRTs, currentTotalSize, batchID)
+				if err != nil {
+					return errors.Trace(err)
+				}
+				currentTotalSize = 0
+				return nil
+			}
+
+			// lowerBound and upperBound are both closed intervals.
+			iter := w.sortEngine.FetchByTable(task.tableID, task.lowerBound, upperBound)
+			for {
+				e, pos, err := iter.Next()
+				if err != nil {
+					return errors.Trace(err)
+				}
+				// There is no more data.
+				if e == nil {
+					// This means that we append all the events to the table sink.
+					// But we have not updated the resolved ts.
+					// Because we do not reach the maxUpdateIntervalSize.
+					if currentTotalSize != 0 {
+						if err := advanceTableSinkAndResetCurrentSize(e); err != nil {
+							return errors.Trace(err)
+						}
+					}
+					break
+				}
+				for availableMem-e.Row.ApproximateBytes() < 0 {
+					if !w.splitTxn {
+						w.memQuota.forceAcquire(defaultRequestMemSize)
+					} else {
+						err := w.memQuota.blockAcquire(defaultRequestMemSize)
```

Comment:
How about doing a flush before this line?

Reply:
I think we should not flush it. Because we only if the split txn enabled, we only flush it when we reach a batch(50MB).